### ISSUE TITLE
Future provider memory leak

### ISSUE THREAD

**Issue by Colman** [rrousselGit/riverpod#3522]
**Describe the bug**
Listeners are added when they shouldn't be when using future providers. This leads to a memory leak since the listeners are never removed.

**Dependencies**
```yaml
flutter_riverpod: 2.5.1
riverpod_annotation: 2.3.5
riverpod_generator: 2.4.0
riverpod_lint: 2.3.10
```

**Code**
```dart
@riverpod
Future<int> a(ARef ref) async {
  print('buildA');
  ref.onAddListener(() {
    print('addA');
  });
  ref.onRemoveListener(() {
    print('removeA');
  });
  ref.onCancel(() {
    print('cancelA');
  });
  await Future.delayed(const Duration(seconds: 5));
  await ref.watch(bProvider.future);
  return 5;
}

@riverpod
Future<int> b(BRef ref) async {
  print('buildB');
  ref.onAddListener(() {
    print('addB');
  });
  ref.onRemoveListener(() {
    print('removeB');
  });
  ref.onCancel(() {
    print('cancelB');
  });
  return 6;
}

class SomeScreen extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    ref.watch(aProvider);
    return Container();
  }
}
```

If you display `SomeScreen`, then click back within 5 seconds (the delay in `aProvider`), the following is printed:
```
buildA
addA
removeA
cancelA
buildB
addB
```

A listener for `bProvider` is added but never removed even though nothing is listening to `aProvider` anymore. Shouldn't `aProvider` check its own aliveness before adding a listener to `bProvider` on the line `ref.watch(bProvider.future)`?

**Expected behavior**
The listener on `bProvider` should never be added

**Comment by Colman** [comment]
I guess I just expected both `A` and `B` to be disposed when I leave the page. I expected the state of `B` to be reset when I go back to the page, but it remains. Why should a disposed resource `A` keep its dependencies `B` alive?

I find it weird that an unused auto-dispose provider can still be listening to something. This fundamentally causes a memory leak because if you:
1. Navigate to `SomeScreen`
2. Go back
3. Repeat steps 1-2 many times

In the end, you'll only have one `SomeScreen` mounted, so `A` will end up with one listener. `B` will have 1 useful listener and many other "ghost" listeners that are never cleared. So if those steps are repeated, you end up with 1 million listeners on b even though only one of them is for a resource that isn't disposed? There's no way that this can be expected behavior.

Why doesn't the watch method in `A` check if `A` is used/not disposed before adding the listener to `B`?

**Comment by Colman** [comment]
My project relies on this task being fixed urgently. If I were to make a PR and it's merged successfully, could it be a part of a minor release, or would it have to wait until 3.0 anyway?

**Comment by rrousselGit** [comment]
It depends on the changes involved. But if it has to be a 3.0.0 fix, you could always stick to using your fork in the meantime

**Comment by Colman** [comment]
I'm not exactly sure how to fix this. If provider `a` in my example is disposed since we left the screen, should it be allowed to watch anything? Or should `A` wait until it has no listeners **and** its build is completed before disposing?

**Comment by mattermoran** [comment]
@rrousselGit I managed to fix it by simply checking the `_mounted` status in ref functions and throwing error if not but I don't know if it could cause any other issues?
here's the pr https://github.com/rrousselGit/riverpod/pull/3598

**Comment by AhmedLSayed9** [comment]
I've got the same bug (which was hard to catch).

A temporary solution is to update aProvider to watch the future before the await:
```dart
@riverpod
Future<int> a(ARef ref) async {
  print('buildA');
  ref.onAddListener(() {
    print('addA');
  });
  ref.onRemoveListener(() {
    print('removeA');
  });
  ref.onCancel(() {
    print('cancelA');
  });
  final bFuture = ref.watch(bProvider.future);
  await Future.delayed(const Duration(seconds: 5));
  await bFuture;
  return 5;
}
```

**Comment by mattermoran** [comment]
@AhmedLSayed9 that kinda works but you would end up making that network call even if provider is disposed and there's no need. also it gets quite messy the more futures you have in your provider. 
but if you ok with waiting all those futures even when disposed you might want to consider `keepAlive` link instead in this case you only have 2 extra lines no matter how many futures you have
```dart
@riverpod
Future<int> a(ARef ref) async {
  print('buildA');
  ref.onAddListener(() {
    print('addA');
  });
  ref.onRemoveListener(() {
    print('removeA');
  });
  ref.onCancel(() {
    print('cancelA');
  });
  
  final keepAliveLink = ref.keepAlive(); // make it non disposable
  
  await Future.delayed(const Duration(seconds: 5));
  await ref.watch(bProvider.future);
  
  keepAliveLink.close(); // release
  
  return 5;
}
```

**Comment by AhmedLSayed9** [comment]
> you would end up making that network call even if provider is disposed

This would occur in both cases. I'd use `CancelToken` for that matter anyway.

The `KeepAliveLink` will prevent `aProvider` from being disposed if `bProvider` throws an error.

**Comment by AhmedLSayed9** [comment]
> you would end up making that network call even if provider is disposed

This would occur in both cases. I'd use `CancelToken` for that matter anyway.

> you might want to consider keepAlive link instead

The `KeepAliveLink` will prevent `aProvider` from being disposed if `bProvider` throws an error.

**Comment by mattermoran** [comment]
> This would occur in both cases. I'd use CancelToken for that matter anyway.

Yup that's why I said if you ok with that. Some requests unfortunately can't be canceled e.g supabase db calls, etc

> The KeepAliveLink will prevent aProvider from being disposed if bProvider throws an error.

Fair enough. You'd want to have a wrapper function that would handle with try/catch/finally which gets quite ugly regardless.
 
The proper solution is to just throw when trying to use watch/listen of disposed refs which is coming in v3. For now we may just have to stick to a fork

**Comment by Colman** [comment]
> The proper solution is to just throw when trying to use watch/listen of disposed refs which is coming in v3. For now we may just have to stick to a fork

It seems weird to me that a framework exception like that would be thrown in a provider. The programmer didn't really do anything wrong and shouldn't have to handle such exception. If they have a try/catch block in there, that new exception will change the control flow of the provider. Why should having no listeners on the provider change control flow in the build method like this?

It seems more natural to me to have a "disposing" state along with a "disposed" state of a provider. If an auto dispose future provider loses its last listener, it should go into a "disposing" state and remain in that state until it's current build is completed. If a different provider tries to listen to a "disposing" provider, Riverpod would just make a new instance of the provider and let the build method of the "disposing" instance run out. If a "disposing" provider tries to use its ref to watch another provider, Riverpod should let it. Then once the "disposing" provider is done building, it would dispose and clear all of its dependencies.

Normal async functions in dart can't be cancelled, you can only ignore their result. It seems weird to me to fight that behavior by throwing exceptions.

**Comment by lucavenir** [comment]
Hello 👋🏼
I can reproduce this issue on v 2.6.1, but in 3.0 it's been fixed. Indeed, v3 introduces `ref.mounted`.

<details><summary>self-enclosing repro test</summary>

```dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('test name', () async {
    const delay = Duration(milliseconds: 500);
    var baseListeners = 0;
    final baseProvider = FutureProvider.autoDispose<int>((ref) {
      ref.onAddListener(() {
        baseListeners++;
      });
      ref.onRemoveListener(() {
        baseListeners--;
      });
      return Future.value(1);
    }, name: "base");
    var plusOneListeners = 0;
    final plusOneProvider = FutureProvider.autoDispose<int>((ref) async {
      ref.onAddListener(() {
        plusOneListeners++;
      });
      ref.onRemoveListener(() {
        plusOneListeners--;
      });
      await Future.delayed(delay);
      final two = await ref.watch(baseProvider.future);
      return two + 1;
    }, name: "plusOne");
    final container = ProviderContainer(observers: []);
    final sub = container.listen(plusOneProvider, (previous, next) {});

    await Future.delayed(delay * 0.90);
    sub.close();
    await Future.delayed(delay * 0.10);

    expect(plusOneListeners, isZero);  // passes
    expect(baseListeners, isZero);  // fails
  });
}
```
</details>

If you run the above test with 3.0, it passes.
Furthermore, you'd notice a nice thrown error explaining a `ref` has been used after disposal.

**Comment by rrousselGit** [comment]
Fixed, so closing
